### ISSUE TITLE
make the C++ code conform to the style guide

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#2902]
Short version:

If you are working on a change that isn't merged please add a comment that states which files in the *src/* directory you are working on. I intend to reformat source not mentioned in this issue two weeks after opening this issue. 

Long version:

Now that there is a trivial way (`make style` and `make style-all`) to ensure code being changed adheres to the style guide it is time to reformat the existing code once again. This was most recently done in 2014 (via two commits by @ridiculousfish). Even after that cleanup the code still deviated significantly from the published style because the tools used to reformat the code (e.g., `astyle`) were not very good at accomplishing the goal. They were also not automated and subsequent pull-request reviews did not enforce adherence to the style guide.

Note that reformatting the code will also eliminate a large number of lint warnings reported by `make lint-all` which will make it easier for contributors to notice if their change introduces new "lint".

The question is how to proceed given that the existing code deviates greatly from the style guide. I propose we start by reformatting every file that no one is working on. Everyone with a change that is not merged should add a comment that states which files in the *src/* directory you are working on. We will avoid reformatting those files in the initial set of changes to update the code to conform to the style guide. This promise is valid for two weeks from the date this issue is opened.

**Comment by floam** [comment]
So if `make style` will be applied in two weeks, isn't a better solution that I just run `make style` now for something that will likely be merged after then? It's not a hardship the project should have to wait on.

**Comment by floam** [comment]
But thanks for the notice... I'm working on `builtin_string.cpp` in one branch, `env_universal_common.cpp`in another, and another has changes to: `fallback.h`, `output.cpp`.

Don't worry about improving those files though.

**Comment by zanchey** [comment]
I'll cut the 2.3.0 integration branch beforehand.

**Comment by krader1961** [comment]
> isn't a better solution that I just run make style now for something that will likely be merged after then?

The, admittedly fuzzy, idea is to minimize disruption to work in progress that others might have in there local repo. If you restyle a file you're going to change and someone else also has unmerged changes for the same file it will cause them a lot of of merge conflicts. At some point they'll have to deal with that pain but I'm hoping, with a little care, to limit that to just a couple of really old changes. Ridiculousfish in particular has a couple of old branches which aren't ready to merge to the master branch, and may never be ready, but he doesn't want to lose that work.

As for who actually creates the restyle commits it doesn't really matter as long as the commit contains nothing by style cleanup. We do not want to mix substantive changes with lint and style cleanup changes.

> I'll cut the 2.3.0 integration branch beforehand.

Good point. If that hasn't happened in the next two weeks we'll postpone this cleanup. The deadline is simply to get people mentally prepared for a big change in the codebase and set a firm(ish) date so it actually happens.

**Comment by krader1961** [comment]
FWIW, I have a local branch that modifies *src/history.cpp*. If no one else says they're editing that file it will be the first file I reformat. That way I'll get some first hand experience of the pain this will cause for someone with a local branch that will be affected by this cleanup.

**Comment by krader1961** [comment]
Do you have changes in your local repo not merged into the fish-shell master branch? If you do please add a comment to this issue reporting which files you are modifying but intend to merge soon. Any files not mentioned will be re-formatted (i.e., re-styled) to conform to the style guide a little over one week from today. This will probably cause you great pain when you attempt to reconcile your local changes with the newly style compliant code.

I want to minimize disruptions to changes that people are working on which are not yet merged. However, since I have no visibility into your local git repo I am depending on you to tell me if this cleanup will cause you pain.

**Comment by floam** [comment]
> If no one else says they're editing that file it will be the first file I reformat.

I have changes in a branch. 

https://github.com/floam/fish-shell/commit/ccd78eeecedcddaeec39a9aa18e4a39ae83162d3#diff-9f6ee6fe77c262c9ece7ab89e197ea0b

**Comment by ridiculousfish** [comment]
I have no pending changes to history.cpp

**Comment by krader1961** [comment]
We are a few days away from making radical changes to the formatting of the fish source code. I do not care which style we enforce. Only that there is a consistent style. The code currently does not conform to a specific style. The code as it exists today doesn't just have minor, inconsequential, deviations from the current style guide. Which makes it really hard for a human to understand the code.

I encourage anyone who has strong feelings on this topic to speak now or forever hold their peace. 

Does anyone object to using `clang-format` with the following default configuration:

```
# Use the Google style with these modifications:
#
# 1) lines can be up to 100 chars long rather than 80, and
# 2) use a four space indent rather than two spaces.
#
BasedOnStyle: Google
ColumnLimit: 100
IndentWidth: 4
```

I merged that two weeks ago in commit fd1b7ba52901179c19f5997b650518993d919f52 to close issue #571. No one has voiced any concerns. I encourage you to run `make style-all` and then do a `git diff` or `git multidiff`. If you are unhappy with the result please say so and what alternative formatting you prefer.

**Comment by floam** [comment]
Might it be nicer if we require use of the tool piecemeal, naturally as want to touch areas? Or is there a way to do this without  wreccking the history or otherwise making git blame not work, and blame mode in my IDE not suck?

**Comment by krader1961** [comment]
@floam: This is orthogonal, albeit related to, the lint cleanup.

There is, sadly, no way to do this type of cleanup without "wreccking the history or otherwise making git blame not work" that I am aware of. I refer you to previous changes of this nature. Most notably the one that moved all the source code from the top-level directory to the `src` directory. That is why I want to ensure most stakeholders are in agreement with the new style and how to enforce it. We do not want to have another major re-formatting of the code a few months, or even a few years, from now.

**Comment by floam** [comment]
I think git blame is preserved pretty well if the new style gets applied to a line of code where there was otherwise some reason to make a change, in the same commit (or upon squash merge by our nannies). 

And the other "is everyone out of history.cpp yet? clear!" problems, and other git history problems. The upside is the code changes to the new style tomorrow if you want it to, I suppose it was decided that it's worth the cost and pain.

**Comment by floam** [comment]
Also, why d we really want to use at the Makefile here to apply the format? `git-clang-format` is so much nice on my system I got it when I installed clang-format. 

Further, very wrong:
```
      clang-format $file > $file.new
            if cmp --quiet $file $file.new
                echo $file was correctly formatted
                rm $file.new
            else
                echo $file was NOT correctly formatted
                mv $file.new $file
```

That will not detect what it says it can.

clang-format outputs a Google-convention-styled, 4 character indented, 100 column wrapped file. Not the only one! There isn't only one! That's why it's best to use it in your txt editor on selected lines, or with git-clang`, which can format a commit, and take the code that actually changed into account when doing checks.. 

That kind of a check string comparison on the output works in go and fish but how that that really be want we want with C++?  

And if this will be something for CI tobe working with (I hope so?)... we will definitely be wanting style failures and passes to be limited to the scope of the actual changes in someone's PR actual commits.  (using git-clang-format).

**Comment by krader1961** [comment]
Thanks for pointing out `git-clang-format`. I hadn't noticed that `brew install clang-format` provided that extension. There's absolutely no reason people shouldn't begin using `git clang-format` immediately and obviously you should configure your editor to auto-format according to the style for the project. I'll update *CONTRIBUTING.md* to include info on both.

Sadly, without concerted cleanup effort too much of the code will remain non-conforming for years to come. Which makes a style guide for the project pretty meaningless. Anyone with an unmerged change to the C++ code should definitely use `git clang-format` rather than `make style` to reformat the code being touched. That way those changes won't need to be reformatted but the change itself won't be harder to review due to having lots of irrelevant changes to the rest of the code.

**Comment by ridiculousfish** [comment]
In case anyone is unaware, running a formatter makes `git blame` annoying but not impossible - you just have to back up to the pre-formatting commit and run it from there. E.g. if the format commit is DEADBEEF you could run `git blame file DEADBEEF^ file.cpp`

**Comment by floam** [comment]
> Anyone with an unmerged change to the C++ code should definitely use git clang-format

How about postpone the upcoming mass-restyle a bit - give the common committers a chance to ameleorate the upcoming merge.

Earlier in this thread where I had questioned why we don't just do the style-changes ourselves I was assuming that we both imagined that would be done with `git-clang-format` - a wrong assumption. 

The article I linked suggested something they did in their organization, which was they picked an easy to remember date for the merge (they were near jan 1, so got lucky),

**Comment by krader1961** [comment]
Okay, I've committed the first style cleanup change to restyle the history module (*history.cpp* and *history.h*). Merging that into my configurable-history branch took more work than usual but was straightforward. I'm going to do the builtin module next since it too is in my configurable-history branch. From the commit comment:

```
Make the history code conform to the new style guide. Every change was
produced by clang-format (e.g., `make style`) with the exception of
comments which were manually reformatted.  That has to be done by hand
since clang-format leaves comments alone other than to reflow comment
lines to get them below the allowed line length.

The total number of lines is reduced by 313 lines (13%) in the two
affected files. Line count is generally a poor metric but in this
case it reflects an increase in information density without a loss in
readability. Furthermore, the standardization of braces, whitespace,
and comment style will make it easier for people to read the code.

This reduces the number of warnings by `make lint` from 168 to 87 (a 48%
decrease). Making it much easier to focus on the substantive lint issues.

Further improvements are possible. For example, many comments are not
very helpful (e.g., they point out the obvious) or provide insufficient
detail. But those are beyond the scope of this change.
```

**Comment by krader1961** [comment]
I just updated the *builtin* module to conform to the style guide. I'm going to do one more tonight, the *autoload* module, since it is very unlikely anyone is working on it. Then I'll let these changes bake for a few days before proceeding with additional cleanup.

Note that I am not removing any Doxygen annotations in any of these commits (other than the `\file` annotation at the top of the module). Thus, if we collectively decide that we want to enforce either Doxygen and/or HeaderDoc annotations the existing ones will still be there as a starting point. It will simply require deciding what our preferred annotation standard will be then adapting the existing comments to conform to that standard (which should be a trivial change).

Please think carefully before you say "sure, let's support code annotations" because if you do then I'm going to expect that every single function, class definition, etc. have the appropriate annotations. Both now and in any future change. Including the annotations only if you feel like it isn't acceptable if they're to have any value. At $dayjob the Python projects I work on require strict adherence to class and method/function annotation standards to document arguments and return values. I would prefer that be the norm for the fish shell. But such annotations are only useful (i.e., trustworthy) if we're hard-nosed about every single change adhering to that requirement. I've seen no evidence that the core fish developers are willing to push back on changes that don't include appropriate source comments let alone amend those changes with the necessary documentation.

I prefer we stop pretending that running Doxygen or Xcode HeaderDoc over the source produces anything close to authoritative documentation if we're not going to actually enforce that standard for documenting the source.

**Comment by krader1961** [comment]
Restyling the *autoload* module resulted in these stats (from the commit comment):

```
Reduces lint errors from 38 to 19 (-50%). Line count from 506 to 426 (-16%).
```

@floam seems to believe the status quo vis-a-vis code annotation is good enough. I do not agree.  I believe that if we want code annotation to be useful it has to be enforced consistently. I would prefer that classes, methods, functions, etc. were annotated. But the current code base is mostly devoid of useful annotations. I think it is preferable that tools like Doxygen and HeaderDoc report nothing at all rather than incomplete and bogus information. If we want those tools to provide useful information then we need to become a lot more hard-nosed (i.e., determined) regarding changes that should include appropriate annotations.

**Comment by floam** [comment]
>  I've seen no evidence that the core fish developers are willing to push back on changes that don't include appropriate source comments let alone amend those changes with the necessary documentation.

Will `//` change that?

At $dayjob your coworkers are not simply fellow volunteers who share an interest - they are professionals that are being handsomely compensated to endure the hard-nosed atmosphere. 

For me, this most recent commit, after our last discussion, seems to simply break some toys in order to prove a point - I suppose for tonight I'll take my ball and go home as there's clearly not much to be gained by arguing this on based on just its own merits.

**Comment by krader1961** [comment]
> For me, this most recent commit, after our last discussion, seems to simply break some toys in order to prove a point...

Not at all. I would prefer that the code have a consistent annotation style regarding arguments and return values. Regardless of whether that conforms to Doxygen, HeaderDoc, or some other documentation standard. My objection is that the existing style is pretty much useless in this regard and therefore should be abandoned because it is not documented and is not enforced. I have no objection to specifying, and more importantly, enforcing a particular style for documenting the source code.

**Comment by krader1961** [comment]
> Will `//` change that?

It will if enough people like yourself complain about the loss of information. Thus forcing a change to the style guide and source code. In particular, forcing the inclusion of the appropriate annotations everywhere rather than just wherever they were incidentally included. Like you, @floam, I would prefer meaningful annotations. The problem is that a) the current code base is extremely inconsistent about this, and b)  nobody is rejecting, or has a good reason to reject, changes that violate that standard.
